packages/identify/src/identify.ts

```diff
@@ -0,0 +1,167 @@
+      event_type: SpecialEventType.IDENTIFY,
+      groups: { ...this._groups },
+      user_properties: this.getUserProperties(),
+      user_id: userId,
+    };
+
+    if (deviceId !== null && deviceId.length > 0) {
+      identifyEvent.device_id = deviceId;
+    }
+
+    return identifyEvent;
+  }
+
+  protected getUserProperties(): IdentifyUserProperties {
+    const userPropertiesCopy: IdentifyUserProperties = {};
+    for (const field of USER_IDENTIFY_OPERATIONS) {
+      if (field in this._properties) {
+        userPropertiesCopy[field] = this._properties[field];
+      }
+    }
+
+    return userPropertiesCopy;
+  }
+
+  public setGroup(groupName: string, groupValue: string): Identify {
+    const isSuccessfulSet = this._safeSet(IdentifyOperation.SET, groupName, groupValue);
+    if (isSuccessfulSet) {
+      this._groups[groupName] = groupValue;
+    }
+
+    return this;
+  }
+
+  public set(property: string, value: ValidPropertyType): Identify {
+    this._safeSet(IdentifyOperation.SET, property, value);
+    return this;
+  }
+
+  public setOnce(property: string, value: ValidPropertyType): Identify {
+    this._safeSet(IdentifyOperation.SET_ONCE, property, value);
+    return this;
+  }
+
+  public append(property: string, value: ValidPropertyType): Identify {
+    this._safeSet(IdentifyOperation.APPEND, property, value);
+    return this;
+  }
+
+  public prepend(property: string, value: ValidPropertyType): Identify {
+    this._safeSet(IdentifyOperation.PREPEND, property, value);
+    return this;
+  }
+
+  public postInsert(property: string, value: ValidPropertyType): Identify {
+    this._safeSet(IdentifyOperation.POSTINSERT, property, value);
+    return this;
+  }
+
+  public preInsert(property: string, value: ValidPropertyType): Identify {
+    this._safeSet(IdentifyOperation.PREINSERT, property, value);
+    return this;
+  }
+
+  public remove(property: string, value: ValidPropertyType): Identify {
+    this._safeSet(IdentifyOperation.REMOVE, property, value);
+    return this;
+  }
+
+  public add(property: string, value: number): Identify {
+    this._safeSet(IdentifyOperation.ADD, property, value);
+    return this;
+  }
+
+  public unset(property: string): Identify {
+    this._safeSet(IdentifyOperation.UNSET, property, UNSET_VALUE);
+    return this;
+  }
+
+  public clearAll(): Identify {
+    // When clear all happens, all properties are unset. Reset the entire object.
+    this._properties = {};
+    this._properties[IdentifyOperation.CLEAR_ALL] = UNSET_VALUE;
+
+    return this;
+  }
+
+  // Returns whether or not this set actually worked.
+  private _safeSet(operation: IdentifyOperation, property: string, value: any): boolean {
+    if (this._validate(operation, property, value)) {
+      if (!(operation in this._properties)) {
+        this._properties[operation] = {};
+      }
+
+      let userPropertyMap: any = this._properties[operation];
+      if (userPropertyMap === undefined) {
```

Comment:
would this ever happen? since you have lines 115-117 above

Reply:
oh! thanks for the callout - this one was the more "typescript friendly" version. I'll delete 115-117